### PR TITLE
[#616] Download Stream generates too many updates on the main thread

### LINKED ISSUE #616: Download Stream generates too many updates on the main thread
This function 

````Swift 

    func compactBlockBatchDownload(
        range: CompactBlockRange,
        batchSize: Int = 100,
        maxRetries: Int = 5
    ) async throws {
````

calls 
````Swift
    notifyProgress(
                    .download(
                        BlockProgress(
                            startHeight: startHeight,
                            targetHeight: targetHeight,
                            progressHeight: range.upperBound
                        )
                    )
                )
````

once per block causing too many updates on the main thread, this being detrimental to performance and energy consumption.

````Swift
func notifyProgress(_ progress: CompactBlockProgress) {
        var userInfo: [AnyHashable: Any] = [:]
        userInfo[CompactBlockProcessorNotificationKey.progress] = progress

        LoggerProxy.debug("progress: \(progress)")
        
        NotificationCenter.default.mainThreadPost(
            name: Notification.Name.blockProcessorUpdated,
            object: self,
            userInfo: userInfo
        )
    }
````

**solution:**
add a parameter to batch progress updates and move the updates to a background thread if possible only for UI implementations to move the updates to the main thread instead

### PR DESCRIPTION
**Pull request by pacu**
Closes #616

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [ ] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [ ] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [ ] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/mater/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage befor and after your changes and when possible, leave it in a better place. [Learn More...](../blob/master/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/master/README.md), [LICENSE.md](../blob/master/LICENSE.md), and [Architecture.md](../blob/master/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

File: Sources/ZcashLightClientKit/Block/Processor/CompactBlockProcessor.swift [hunk 1]
```diff
@@ -635,7 +635,7 @@ public actor CompactBlockProcessor {
 
         LoggerProxy.debug("progress: \(progress)")
         
-        NotificationCenter.default.mainThreadPost(
+        NotificationCenter.default.post(
```
Comment (Chlup) [inline Sources/ZcashLightClientKit/Block/Processor/CompactBlockProcessor.swift:638]: This means that SDK now delivers some notifications on main thread and some on some other threads. Are we ok with this inconsistency? Also I am not sure if this is enough. App eventually must go to main thread after receiving this notification to update UI. So the problem stays it's just moved.

To solve the problem app must implement some throttling mechanism. Can't we simply send this progress update after N blocks or after some time period to make it much simpler to use? 

If we are in hurry with 0.17.0 release we can go with this workaround. But if we are not maybe we can improve this fix.

**Review by Chlup — CHANGES_REQUESTED** [review]

File: Sources/ZcashLightClientKit/Block/Processor/CompactBlockProcessor.swift (hunk 1 above)
Comment (pacu) [reply Sources/ZcashLightClientKit/Block/Processor/CompactBlockProcessor.swift:638]: it's definitely not enough. I've implemented the limiting counterpart on the client. We do have to patch it but we can definitely improve it

Comment (Chlup) [reply Sources/ZcashLightClientKit/Block/Processor/CompactBlockProcessor.swift:638]: Ok. Let's go with this for now.

**Review by Chlup — APPROVED** [review]

File: Sources/ZcashLightClientKit/Block/Processor/CompactBlockProcessor.swift (hunk 1 above)
Comment (Chlup) [reply Sources/ZcashLightClientKit/Block/Processor/CompactBlockProcessor.swift:638]: I just realized. `SDKSynchronizer` subscribes for `processorUpdated` notification. And when it receives it then `SDKSynchronizer` sends notification `synchronizerProgressUpdated` on main thread. So maybe we should send even this notification not on main thread?

**Review by Chlup — CHANGES_REQUESTED** [review]

File: Sources/ZcashLightClientKit/Block/Processor/CompactBlockProcessor.swift (hunk 1 above)
Comment (pacu) [reply Sources/ZcashLightClientKit/Block/Processor/CompactBlockProcessor.swift:638]: Yes, that's a good catch. 

The reference to self is problematic here. 


````Swift
// MARK: notify state
    private func notify(progress: CompactBlockProgress) {
        var userInfo: [AnyHashable: Any] = .init()
        userInfo[NotificationKeys.progress] = progress
        userInfo[NotificationKeys.blockHeight] = progress.progressHeight

        self.status = SyncStatus(progress)
        NotificationCenter.default.mainThreadPost(name: Notification.Name.synchronizerProgressUpdated, object: self, userInfo: userInfo)
    }
````

Comment (Chlup) [reply Sources/ZcashLightClientKit/Block/Processor/CompactBlockProcessor.swift:638]: Do you plan to change this also in this PR? Or we are going without it?

Comment (Chlup) [reply Sources/ZcashLightClientKit/Block/Processor/CompactBlockProcessor.swift:638]: Do you plan to change this in this PR? Or are we going without it?

**Comment by pacu** [comment]
I'm testing this with darksidewalletd and noticed some errors and warnings  of SDK being started when it's syncing.

I'll move this to draft

**Comment by pacu** [comment]
Closing since @Chlup will take over this

**Pull request closed without merging**